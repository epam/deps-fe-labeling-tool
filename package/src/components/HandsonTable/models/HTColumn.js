
import PropTypes from 'prop-types'

class HTColumn {
  constructor (data) {
    this.data = data
  }
}

const htColumnShape = PropTypes.shape({
  data: PropTypes.string.isRequired
})

export {
  HTColumn,
  htColumnShape
}
