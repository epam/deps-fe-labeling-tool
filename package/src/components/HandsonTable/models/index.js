
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
