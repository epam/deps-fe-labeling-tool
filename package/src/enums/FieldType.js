const FieldType = {
  CHECKMARK: 'checkmark',
  ENUM: 'enum',
  LIST: 'list',
  PAIR: 'pair',
  STRING: 'string',
  TABLE: 'table',
  DATE: 'date'
}

const RESOURCE_FIELD_TYPE = {
  [FieldType.STRING]: 'String',
  [FieldType.ENUM]: 'Enumeration',
  [FieldType.CHECKMARK]: 'Checkbox',
  [FieldType.LIST]: 'List',
  [FieldType.PAIR]: 'Key Value Pair',
  [FieldType.TABLE]: 'Table',
  [FieldType.DATE]: 'Date'
}

export {
  FieldType,
  RESOURCE_FIELD_TYPE
}
