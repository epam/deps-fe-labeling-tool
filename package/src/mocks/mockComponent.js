const mockComponent = (componentName, namedExport = true, fn = () => { }) => {
  Object.defineProperty(fn, 'name', { value: componentName })
  return !namedExport ? fn : { [componentName]: fn }
}

export {
  mockComponent
}
