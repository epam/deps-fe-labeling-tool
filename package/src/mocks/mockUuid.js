const mockUuid = {
  v4: ((n = 1) => () => `${n++}`)()
}

export {
  mockUuid
}
