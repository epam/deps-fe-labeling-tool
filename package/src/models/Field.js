import PropTypes from 'prop-types'
import { FieldType } from '@/enums/FieldType'
import {
  enumFieldMetaShape,
  listFieldMetaShape,
  pairFieldMetaShape
} from '@/models/FieldMeta'

const PRIMITIVE_FIELD_TYPES = [FieldType.STRING, FieldType.PAIR, FieldType.TABLE]

class Field {
  constructor (code, name, fieldType, fieldMeta, required = false) {
    this.code = code
    this.name = name
    this.fieldType = fieldType
    fieldMeta && (this.fieldMeta = fieldMeta)
    this.required = required
  }

  static hasPair = (fields, fieldCode) => {
    const field = fields.find((f) => f.code === fieldCode)

    if (field.fieldType === FieldType.LIST) {
      return field.fieldMeta.baseType === FieldType.PAIR
    }

    return field.fieldType === FieldType.PAIR
  }

  static isPrimitive = (fields, fieldCode) => {
    const { fieldType } = fields.find((f) => f.code === fieldCode)
    return PRIMITIVE_FIELD_TYPES.includes(fieldType)
  }
}

const fieldShape = PropTypes.shape({
  code: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  fieldType: PropTypes.oneOf(Object.values(FieldType)).isRequired,
  fieldMeta: PropTypes.oneOfType([
    enumFieldMetaShape,
    listFieldMetaShape,
    pairFieldMetaShape
  ]),
  index: PropTypes.number,
  required: PropTypes.bool.isRequired
})

export {
  fieldShape,
  Field
}
