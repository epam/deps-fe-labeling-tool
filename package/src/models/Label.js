import { Validator } from 'jsonschema'
import PropTypes from 'prop-types'
import { v4 as uuidv4 } from 'uuid'
import { Direction } from '@/enums/Rotation'
import { Rectangle } from './Rectangle'

const LABEL_ERROR_MESSAGE = 'Label is invalid'

const LabelType = {
  CHECKMARK: 'checkmark',
  KEY: 'key',
  VALUE: 'value',
  STRING: 'string',
  ENUM: 'enum',
  DATE: 'date',
  UNASSIGNED: 'unassigned'
}

const LABEL_TYPE_NAME = 'label'

const LABEL_SHIFT_COEFFICIENT = 1.01

class Label {
  typeName = LABEL_TYPE_NAME

  constructor (x, y, w, h, fieldCode = '', index, type = LabelType.UNASSIGNED, content = null, meta = {}, confidence) {
    this.uid = uuidv4()
    this.x = +(x.toFixed(12))
    this.y = +(y.toFixed(12))
    this.w = +(w.toFixed(12))
    this.h = +(h.toFixed(12))
    this.fieldCode = fieldCode
    this.index = index
    this.type = type
    this.content = content
    this.meta = meta
    this.confidence = confidence
  }

  static isAssigned = (label) => !!label.fieldCode

  static isValid = (label) => {
    try {
      const validator = new Validator()
      const { valid } = validator.validate(label, labelSchema)
      if (!valid) {
        throw new Error(LABEL_ERROR_MESSAGE)
      }
      return true
    } catch (e) {
      // eslint-disable-next-line no-console
      console.log(e)
      return false
    }
  }

  static fromRectangle = (rectangle) => {
    const label = new Label(
      rectangle.x,
      rectangle.y,
      rectangle.w,
      rectangle.h
    )

    return label
  }

  static toRectangle = (label) => {
    return new Rectangle(
      label.x,
      label.y,
      label.w,
      label.h
    )
  }

  static getName = (label) => {
    const { fieldCode, index, type } = label
    const idx = index !== undefined ? `.${index}` : ''
    return fieldCode && type ? `${fieldCode}${idx}.${type}` : ''
  }

  static scale = (label, scale) => ({
    ...label,
    ...Rectangle.scale(label, scale)
  })

  static areSame = (label1, label2) => (
    label1.fieldCode === label2.fieldCode &&
    label1.type === label2.type &&
    label1.index === label2.index
  )

  static getSameLabel = (label, allLabels) => allLabels.find((l) => Label.areSame(l, label))

  static getSameLabels = (label, allLabels) => allLabels.filter((l) => Label.areSame(l, label))

  static getSameFieldCodeTypeLabels = (label, allLabels) => allLabels.filter(
    (l) => l.fieldCode === label.fieldCode && l.type === label.type
  )

  static duplicate = (label) => new Label(
    label.x,
    label.y,
    label.w,
    label.h,
    undefined,
    undefined,
    undefined,
    label.content,
    { ...label.meta },
    label.confidence
  )

  static shift = (label) => ({
    ...label,
    x: label.x * LABEL_SHIFT_COEFFICIENT,
    y: label.y * LABEL_SHIFT_COEFFICIENT
  })

  static rotate = (label, direction) => {
    const x = direction === Direction.CLOCKWISE ? 1 - (label.y + label.h) : label.y
    const y = direction === Direction.CLOCKWISE ? label.x : 1 - (label.x + label.w)

    return ({
      ...label,
      x,
      y,
      w: label.h,
      h: label.w
    })
  }

  static getStringContent = (label) => {
    if (label.content === null) {
      return ''
    }
    return label.content.toString()
  }
}

const labelShape = PropTypes.shape({
  typeName: PropTypes.string.isRequired,
  uid: PropTypes.string.isRequired,
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired,
  w: PropTypes.number.isRequired,
  h: PropTypes.number.isRequired,
  fieldCode: PropTypes.string.isRequired,
  index: PropTypes.number,
  type: PropTypes.oneOf(
    Object.values(LabelType)
  ).isRequired,
  content: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.bool
  ]),
  meta: PropTypes.objectOf(PropTypes.any),
  confidence: PropTypes.number
})

const labelSchema = {
  type: 'object',
  properties: {
    uid: { type: 'string' },
    x: { type: 'number' },
    y: { type: 'number' },
    w: { type: 'number' },
    h: { type: 'number' },
    fieldCode: { type: 'string' },
    index: { type: 'integer' },
    type: { enum: Object.values(LabelType) },
    content: {
      type: {
        oneOf: [
          {
            type: 'string'
          },
          {
            type: 'boolean'
          },
          {
            type: 'null'
          }
        ]
      }
    },
    confidence: {
      type: {
        oneOf: [
          {
            type: 'number'
          },
          {
            type: 'null'
          }
        ]
      }
    },
    meta: { type: 'object' },
    typeName: { enum: ['label'] }
  },
  additionalProperties: false,
  required: ['x', 'y', 'w', 'h']
}

export {
  LABEL_TYPE_NAME,
  LabelType,
  labelShape,
  labelSchema,
  Label
}
