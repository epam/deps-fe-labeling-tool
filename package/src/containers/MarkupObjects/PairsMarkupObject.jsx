import React, { useCallback } from 'react'
import { includes } from 'lodash' //
import PropTypes from 'prop-types'
import { fieldShape } from '@/models/Field'
import { labelShape, Label } from '@/models/Label'
import { getRelatedLabels } from '@/models/Relation'
import { MarkupObjectHeader } from './MarkupObjectHeader'
import {
  Circle,
  Dot,
  IconWrapper,
  KeyValuePairWrapper,
  KeyValueItem,
  Tail
} from './PairsMarkupObject.styles'
import { PrimitiveMarkupObjects } from './PrimitiveMarkupObjects'

const RESOURCE_BLANK_PAIR = 'No items here'

const BlankPairItem = () => (
  <MarkupObjectHeader
    name={RESOURCE_BLANK_PAIR}
    unassigned
  />
)

const generateKeyValuePair = (markupObjects) => {
  return markupObjects.reduce((result, markupObject) => {
    if (result.some((item) => includes(item?.key, markupObject) || includes(item?.value, markupObject))) {
      return result
    }

    const relations = getRelatedLabels(markupObjects, markupObject)
    const sameLabels = Label.getSameLabels(markupObject, markupObjects)

    if (!relations?.length) {
      result.push({ [markupObject.type]: sameLabels })
    } else {
      result.push({
        [markupObject.type]: sameLabels,
        [relations[0].type]: relations
      })
    }

    return result
  }, [])
}

const PairsMarkupObject = ({
  markupObjects,
  field
}) => {
  const renderMarkupObject = useCallback((objects) => {
    if (!objects) {
      return <BlankPairItem />
    }

    return (
      <PrimitiveMarkupObjects
        markupObjects={objects}
        field={field}
      />
    )
  }, [field])

  return generateKeyValuePair(markupObjects).map((entity, i) => (
    <KeyValuePairWrapper key={i}>
      <IconWrapper>
        <Circle>
          <Dot />
        </Circle>
        <Tail />
      </IconWrapper>
      <KeyValueItem>
        {renderMarkupObject(entity.key)}
        {renderMarkupObject(entity.value)}
      </KeyValueItem>
    </KeyValuePairWrapper>
  ))
}

PairsMarkupObject.propTypes = {
  field: fieldShape.isRequired,
  markupObjects: PropTypes.arrayOf(labelShape).isRequired
}

export {
  PairsMarkupObject
}
