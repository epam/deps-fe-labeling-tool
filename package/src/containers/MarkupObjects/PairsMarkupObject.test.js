import { mockComponent } from '@/mocks/mockComponent'
import { mockUuid } from '@/mocks/mockUuid'
import React from 'react'
import { shallow } from 'enzyme'
import { FieldType } from '@/enums/FieldType'
import { Field } from '@/models/Field'
import { LabelType, Label } from '@/models/Label'
import { PairsMarkupObject } from './PairsMarkupObject'

jest.mock('uuid', () => mockUuid)
jest.mock('@/models/Markup', () => ({
  ...jest.requireActual('@/models/Markup'),
  Markup: {
    getPageByObjectUid: jest.fn(() => 1)
  }
}))
jest.mock('./MarkupObjectHeader', () => mockComponent('MarkupObjectHeader'))

describe('Component: PairsMarkupObject', () => {
  describe('component', () => {
    let defaultProps, wrapper

    beforeEach(() => {
      defaultProps = {
        field: new Field('testPairCode', 'test Pair Name', FieldType.PAIR),
        markupObjects: [
          new Label(1, 2, 3, 4, 'testPairCode', undefined, LabelType.KEY, ''),
          new Label(1, 2, 3, 4, 'testPairCode', undefined, LabelType.VALUE, '')
        ]
      }

      wrapper = shallow(<PairsMarkupObject {...defaultProps} />)
    })

    it('should render correct layout', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})
