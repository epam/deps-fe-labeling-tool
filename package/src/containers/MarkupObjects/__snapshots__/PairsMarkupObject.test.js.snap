// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: PairsMarkupObject component should render correct layout 1`] = `
<PairsMarkupObjectstyles__KeyValuePairWrapper
  key="0"
>
  <PairsMarkupObjectstyles__IconWrapper>
    <PairsMarkupObjectstyles__Circle>
      <PairsMarkupObjectstyles__Dot />
    </PairsMarkupObjectstyles__Circle>
    <PairsMarkupObjectstyles__Tail />
  </PairsMarkupObjectstyles__IconWrapper>
  <PairsMarkupObjectstyles__KeyValueItem>
    <Connect(PrimitiveMarkupObjects)
      field={
        Field {
          "code": "testPairCode",
          "fieldType": "pair",
          "name": "test Pair Name",
          "required": false,
        }
      }
      markupObjects={
        Array [
          Label {
            "confidence": undefined,
            "content": "",
            "fieldCode": "testPairCode",
            "h": 4,
            "index": undefined,
            "meta": Object {},
            "type": "key",
            "typeName": "label",
            "uid": "1",
            "w": 3,
            "x": 1,
            "y": 2,
          },
        ]
      }
    />
    <Connect(PrimitiveMarkupObjects)
      field={
        Field {
          "code": "testPairCode",
          "fieldType": "pair",
          "name": "test Pair Name",
          "required": false,
        }
      }
      markupObjects={
        Array [
          Label {
            "confidence": undefined,
            "content": "",
            "fieldCode": "testPairCode",
            "h": 4,
            "index": undefined,
            "meta": Object {},
            "type": "value",
            "typeName": "label",
            "uid": "2",
            "w": 3,
            "x": 1,
            "y": 2,
          },
        ]
      }
    />
  </PairsMarkupObjectstyles__KeyValueItem>
</PairsMarkupObjectstyles__KeyValuePairWrapper>
`;
